File: src/test/java/net/dean/jraw/test/HttpRequestTest.java
package net.dean.jraw.test;

import net.dean.jraw.Endpoints;
import net.dean.jraw.http.BasicAuthData;
import net.dean.jraw.http.HttpRequest;
import org.testng.annotations.Test;

import static org.testng.Assert.assertEquals;

public class HttpRequestTest {
    @Test
    public void testMinimal() {
        compare("http://example.com/", new HttpRequest.Builder()
                .host("example.com")
                .build());
    }

    @Test
    public void testBasic() {
        compare("https://example.com/hello", new HttpRequest.Builder()
                .get()
                .https(true)
                .host("example.com")
                .path("/hello")
                .build());
    }

    @Test
    public void testPathParams() {
        compare("http://example.com/id/foo", new HttpRequest.Builder()
                .get()
                .https(false)
                .host("example.com")
                .path("/id/{id}", "foo")
                .build());
    }

    @Test
    public void testEndpointPathParams() {
        compare("https://example.com/api/multi/foo/r/bar", new HttpRequest.Builder()
                .get()
                .https(true)
                .host("example.com")
                .endpoint(Endpoints.MULTI_MULTIPATH_R_SRNAME_DELETE, "foo", "bar")
                .build());
    }

    @Test
    public void testNullPath() {
        compare("https://example.com/", new HttpRequest.Builder()
                .get()
                .https(true)
                .host("example.com")
                .path(null)
                .build());
    }

    @Test(expectedExceptions = IllegalArgumentException.class)
    public void testMissingHost() {
        new HttpRequest.Builder().build();
    }

    @Test(expectedExceptions = IllegalArgumentException.class)
    public void testRequestWithBasicAuthNoHttps() {
        new HttpRequest.Builder()
                .host("example.com")
                .basicAuth(new BasicAuthData("foo", "bar"))
                .https(false)
                .build();
    }

    private void compare(String expectedUrl, HttpRequest actual) {
        assertEquals(actual.getUrl().toExternalForm(), expectedUrl);
    }
}

File: src/test/kotlin/net/dean/jraw/test/HttpRequestTest.kt
package net.dean.jraw.test

import com.winterbe.expekt.should
import net.dean.jraw.http.HttpRequest
import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.it
import java.lang.AssertionError

class HttpRequestTest: Spek({
    it("should let us use url() only") {
        val url = "https://google.com"
        HttpRequest.Builder()
            .get()
            .url(url)
            .build().url.should.equal(url)
    }

    it("should let us use secure(), host(), path(), and query() instead of url()") {
        HttpRequest.Builder()
            .get()
            .secure(true)
            .host("google.com")
            .path("/search")
            .query(mapOf("q" to "hello"))
            .build().url.should.equal("https://google.com/search?q=hello")
    }

    it("should throw an error if we leave the host out") {
        try {
            HttpRequest.Builder()
                .path("/foo")
                .build()

            throw AssertionError("Expecting an IllegalArgumentException")
        } catch (ex: Exception) {
            if (ex !is IllegalArgumentException) {
                // We were expecting an IllegalArgumentException
                throw AssertionError("Expecting an IllegalArgumentException")
            }
        }
    }

    it("should handle paths without a leading slash") {
        HttpRequest.Builder()
            .secure(true)
            .host("google.com")
            .path("search")
            .build().url.should.equal("https://google.com/search")
    }

    it("should handle path parameters") {
        HttpRequest.Builder()
            .host("github.com")
            .path("/{user}/{repo}", "JetBrains", "kotlin")
            .build().url.should.equal("https://github.com/JetBrains/kotlin")
    }
})
